src/classes/view/base.vala

```diff
@@ -40,45 +41,234 @@ namespace pdfpc {
          * Renderer to be used for rendering the slides
          */
-        protected Renderer.Base renderer;
+        protected Renderer.Pdf renderer;
+
+        /**
+         * Return the used renderer object
+         */
+        public Renderer.Pdf get_renderer() {
+            return this.renderer;
+        }
+
+        /**
+         * Signal emitted every time a precached slide has been created
+         *
+         * This signal should be emitted slide_count number of times during a
+         * precaching cylce.
+         */
+        public signal void slide_prerendered();
+
+        /**
+         * Signal emitted when the precaching cycle is complete
+         */
+        public signal void prerendering_completed();
+
+        /**
+         * Signal emitted when the precaching cycle just started
+         */
+        public signal void prerendering_started();
+
+        /**
+         * The currently displayed slide
+         */
+        protected int current_slide_number;
+
+        /**
+         * The surface containing the currently shown slide
+         */
+        protected Cairo.ImageSurface current_slide;
+
+        /**
+         * The number of slides in the presentation
+         */
+        protected int n_slides;
+
+        protected int slide_limit;
+
+        /**
+         * List to store all associated behaviours
+         */
+        protected GLib.List<View.Behaviour.Base> behaviours = new GLib.List<View.Behaviour.Base>();
 
         /**
          * Base constructor taking the renderer to use as an argument
          */
-        protected Base( Renderer.Base renderer ) {
+        public Base(Renderer.Pdf renderer) {
             this.renderer = renderer;
             this.set_size_request(renderer.width, renderer.height);
+
+            this.current_slide_number = 0;
+
+            this.n_slides = (int) renderer.metadata.get_slide_count();
+            this.slide_limit = this.n_slides + 1;
+
+            // Render the initial page on first realization.
+            this.add_events(Gdk.EventMask.STRUCTURE_MASK);
+            this.realize.connect(() => {
+                 try {
+                     this.display( this.current_slide_number );
+                 } catch( Renderer.RenderError e ) {
+                     // There should always be a page 0 but you never know.
+                     error("Could not render initial page %d: %s",
+                         this.current_slide_number, e.message);
+                 }
+
+                 // Start the prerender cycle if the renderer supports caching
+                 // and the used cache engine allows prerendering.
+                 // Executing the cycle here to ensure it is executed within the
+                 // Gtk event loop. If it is not proper Gdk thread handling is
+                 // impossible.
+                 var caching_renderer = this.renderer as Renderer.Pdf;
+                 if (caching_renderer != null && caching_renderer.cache != null &&
```

Comment:
It is guaranteed that `this.renderer` is a `Renderer.Pdf` instance, the casting and first condition within the `if` should no longer be required.